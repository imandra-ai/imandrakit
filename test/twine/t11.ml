type bar =
  | Bar1
  | Bar2
[@@deriving eq, show, twine]

type event =
  | E_bar of bar
  | E_prev of event Imandrakit_twine.offset_for
[@@deriving show { with_path = false }, twine]

type foo = { evs: event list } [@@deriving show { with_path = false }, twine]

let encoded : string =
  let enc = Imandrakit_twine.Encode.create () in
  let ev0 =
    Imandrakit_twine.Encode.write_offset_for enc event_to_twine (E_bar Bar1)
  in
  let ev1 =
    Imandrakit_twine.Encode.write_offset_for enc event_to_twine (E_bar Bar2)
  in
  let top = foo_to_twine enc { evs = [ E_prev ev0; E_prev ev1 ] } in
  Imandrakit_twine.Encode.finalize_copy enc ~entrypoint:top

let () =
  Format.printf "serialized:@.%s@." @@ Imandrakit_twine.Dump.dump_string encoded
