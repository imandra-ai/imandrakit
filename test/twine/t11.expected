serialized:
[0x0]: C_0(C_0)
[0x2]: C_0(C_1)
[0x4]: C_1(ref(0x0))
[0x6]: C_1(ref(0x2))
[0x8]: [@0x4, @0x6] (len=2)
[0xb]: [@0x8] (len=1)

